src/libraries/System.Linq.AsyncEnumerable/src/System/Linq/Empty.cs

```diff
@@ -16,7 +16,11 @@ public static partial class AsyncEnumerable
         /// <returns>An empty <see cref="IAsyncEnumerable{T}"/> whose type argument is <typeparamref name="TResult"/>.</returns>
         public static IAsyncEnumerable<TResult> Empty<TResult>() => EmptyAsyncEnumerable<TResult>.Instance;
 
-        private sealed class EmptyAsyncEnumerable<TResult> : IAsyncEnumerable<TResult>, IAsyncEnumerator<TResult>
+        /// <summary>Determines whether <paramref name="source"/> is known to be an always-empty enumerable.</summary>
+        private static bool IsEmpty<TResult>(this IAsyncEnumerable<TResult> source) => source is EmptyAsyncEnumerable<TResult>;
```

Comment:
If we ever add collection expressions to `IAE<T>` it would be awkward to reconcile this check with `[]` instances.

Reply:
Shouldn't we be doing reference comparison here as well? Last time I checked running a runtime type test against a generic was relatively speaking expensive.

Reply:
> Shouldn't we be doing reference comparison here as well

That's what I was saying we can change this to.